File: app/src/main/java/com/github/shadowsocks/job/AclSyncJob.java
package com.github.shadowsocks.job;

import androidx.annotation.NonNull;

import com.evernote.android.job.Job;
import com.evernote.android.job.JobRequest;
import com.github.shadowsocks.ShadowsocksApplication;
import com.github.shadowsocks.utils.IOUtils;
import com.github.shadowsocks.utils.VayLog;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.concurrent.TimeUnit;

/**
 * @author Mygod
 */
public class AclSyncJob extends Job {

    public static final String TAG = AclSyncJob.class.getSimpleName();

    private final String route;

    public AclSyncJob(String route) {
        this.route = route;
    }

    public static int schedule(String route) {
        return new JobRequest.Builder(AclSyncJob.TAG + ':' + route)
                .setExecutionWindow(1, TimeUnit.DAYS.toMillis(28))
                .setRequirementsEnforced(true)
                .setRequiredNetworkType(JobRequest.NetworkType.UNMETERED)
                .setRequiresCharging(true)
                .setUpdateCurrent(true)
                .build().schedule();
    }

    @NonNull
    @Override
    protected Result onRunJob(Params params) {
        String filename = route + ".acl";
        InputStream is = null;
        try {
            if ("self".equals(route)) {
                // noinspection JavaAccessorMethodCalledAsEmptyParen
                is = new URL("https://raw.githubusercontent.com/HMBSbige/Text_Translation/master/ShadowsocksR/" + filename).openConnection().getInputStream();
                IOUtils.writeString(ShadowsocksApplication.app.getApplicationInfo().dataDir + '/' + filename, IOUtils.readString(is));
            }
            return Result.SUCCESS;
        } catch (IOException e) {
            VayLog.e(TAG, "onRunJob", e);
            ShadowsocksApplication.app.track(e);
            return Result.RESCHEDULE;
        } catch (Exception e) {
            // unknown failures, probably shouldn't retry
            VayLog.e(TAG, "onRunJob", e);
            ShadowsocksApplication.app.track(e);
            return Result.FAILURE;
        } finally {
            try {
                if (is != null) {
                    is.close();
                }
            } catch (IOException e) {
                VayLog.e(TAG, "onRunJob", e);
                ShadowsocksApplication.app.track(e);
            }
        }
    }
}

File: app/src/main/java/com/github/shadowsocks/job/AclSyncJob.kt
package com.github.shadowsocks.job

import com.evernote.android.job.*
import com.github.shadowsocks.*
import com.github.shadowsocks.utils.*
import java.io.*
import java.net.*
import java.util.concurrent.*

class AclSyncJob(private val route: String) : Job()
{
	override fun onRunJob(params: Params): Result
	{
		val filename = "$route.acl"
		var `is`: InputStream? = null
		try
		{
			if ("self" == route)
			{

				`is` = URL("https://raw.githubusercontent.com/HMBSbige/Text_Translation/master/ShadowsocksR/$filename").openConnection()
					.getInputStream()
				IOUtils.writeString(ShadowsocksApplication.app.applicationInfo.dataDir + '/'.toString() + filename, IOUtils.readString(`is`!!))
			}
			return Result.SUCCESS
		}
		catch (e: IOException)
		{
			VayLog.e(TAG, "onRunJob", e)
			ShadowsocksApplication.app.track(e)
			return Result.RESCHEDULE
		}
		catch (e: Exception)
		{
			// unknown failures, probably shouldn't retry
			VayLog.e(TAG, "onRunJob", e)
			ShadowsocksApplication.app.track(e)
			return Result.FAILURE
		}
		finally
		{
			try
			{
				`is`?.close()
			}
			catch (e: IOException)
			{
				VayLog.e(TAG, "onRunJob", e)
				ShadowsocksApplication.app.track(e)
			}

		}
	}

	companion object
	{
		const val TAG = "AclSyncJob"

		fun schedule(route: String): Int
		{
			return JobRequest.Builder("$TAG:$route")
				.setExecutionWindow(1, TimeUnit.DAYS.toMillis(28))
				.setRequirementsEnforced(true)
				.setRequiredNetworkType(JobRequest.NetworkType.UNMETERED)
				.setRequiresCharging(true)
				.setUpdateCurrent(true)
				.build()
				.schedule()
		}
	}
}
